syntax = "proto3";

package devicefleetmanager;

service DeviceFleetManagerService {
  rpc RegisterDevice (DeviceRequest) returns (DeviceResponse);
  rpc ListDevices (Empty) returns (DeviceList);
  rpc SetDeviceStatus (StatusRequest) returns (DeviceResponse);
  rpc GetDeviceInfo (DeviceID) returns (DeviceRequest);
  rpc InitiateAction (ActionRequest) returns (ActionResponse);
  rpc GetActionStatus (ActionID) returns (ActionStatusResponse);
}

message Empty {}

message DeviceID {
  int32 id = 1;
}

message DeviceRequest {
  int32 id = 1;
  string name = 2;
  int32 status = 3; //Mapped to C++ enum
}

message DeviceResponse {
  bool success = 1;
  string message = 2;
}

message DeviceList {
  repeated DeviceRequest devices = 1;
}

message StatusRequest {
  int32 id = 1;
  int32 status = 2;
}

message ActionRequest {
  int32 device_id = 1;
  string action_type = 2;
  string params = 3;
}

message ActionResponse {
  int32 action_id = 1;
}

message ActionID {
  int32 id = 1;
}

message ActionStatusResponse {
  int32 status = 1; // 0:PENDING, 1:RUNNING, 2:COMPLETED, 3:FAILED
}

